src/ucp/dt/datatype_iter.inl

```diff
@@ -334,33 +355,70 @@ ucp_datatype_iter_is_end(const ucp_datatype_iter_t *dt_iter)
     return dt_iter->offset == dt_iter->length;
 }
 
+static UCS_F_ALWAYS_INLINE ucs_status_t ucp_datatype_iter_mem_reg_internal(
+        ucp_context_h context, ucp_datatype_iter_t *dt_iter,
+        ucp_md_map_t md_map, void *buffer, size_t length, unsigned uct_flags,
+        ucp_dt_reg_t *dt_reg)
+{
+    ucs_status_t status;
+
+    status = ucp_mem_rereg_mds(context, md_map, buffer, length, uct_flags, NULL,
+                               (ucs_memory_type_t)dt_iter->mem_info.type, NULL,
+                               dt_reg->memh, &dt_reg->md_map);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    /* We expect the registration to happen on all desired memory domains,
+     * since subsequent access to the iterator will use 'memh_index' which
+     * assumes the md_map is as expected.
+     */
+    ucs_assertv((length == 0) || (dt_reg->md_map == md_map),
+                "reg.md_map=0x%" PRIx64 " md_map=0x%" PRIx64, dt_reg->md_map,
+                md_map);
+    return UCS_OK;
+}
+
 /*
  * Register memory and update iterator state
  */
-static UCS_F_ALWAYS_INLINE ucs_status_t
-ucp_datatype_iter_mem_reg(ucp_context_h context, ucp_datatype_iter_t *dt_iter,
-                          ucp_md_map_t md_map, unsigned uct_flags)
+static UCS_F_ALWAYS_INLINE ucs_status_t ucp_datatype_iter_mem_reg(
+        ucp_context_h context, ucp_datatype_iter_t *dt_iter,
+        ucp_md_map_t md_map, unsigned uct_flags, unsigned dt_mask)
```

Comment:
like above?

Reply:
btw, why do we need a mask of DTs, what is use case? looks like it's always being treated as a single dt

Reply:
it will be used by RNDV protocols, and also used by ucp_proto_eager_tag_offload_zcopy_progress->ucp_proto_zcopy_single_progress